File: 0493-reverse-pairs/0493-reverse-pairs.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var reversePairs = function(nums) {
    let ans = 0;
    function mergeSort(nums,si,ei){

    if(si >= ei) {
        return;
    }

    let mid =Math.floor(si+(ei-si)/2);


    mergeSort(nums,si,mid);
    mergeSort(nums,mid+1,ei);
    countPairs(nums, si, mid, ei);
    merge(nums,si,mid,ei);
   
}
   function countPairs(nums, si, mid, ei){
        let j = mid+1;
        for(let i = si; i <= mid;i++){
            while(j <= ei && nums[i]> 2*nums[j]){
                j++;
            }
            ans += j-(mid+1);
        }
 }

function merge(nums,si,mid,ei){
    let i = si;
    let j =mid+1;
    let arr = [];
    while(i <= mid && j <= ei){
        if(nums[i] < nums[j]){
           
           arr.push(nums[i]);
           i++;
        }else {
            arr.push(nums[j]);
            j++;
        }
    }
    while(i <=mid){
        arr.push(nums[i]);
        i++;
    }
     while(j <= ei){
        arr.push(nums[j]);
        j++;
    }
     for (let k = si; k <= ei; k++) {
            nums[k] = arr[k - si];
        }
  }
    mergeSort(nums,0,nums.length-1);
    return ans;
}

File: 0493-reverse-pairs/0493-reverse-pairs.java
class Solution {
    public int reversePairs(int[] nums) {
        if(nums.length <= 1){
            return 0;
        }
        return mergeSort(nums,0,nums.length-1);
    }
    public static int mergeSort(int[] arr, int si, int ei){
        if(si >= ei){
            return 0;
        }
        int count = 0;
        int mid = si+(ei-si)/2;
        count += mergeSort(arr,si,mid);
        count += mergeSort(arr,mid+1,ei);
        count += merge(arr,si,mid,ei);
        return count;
    }
    public static int merge(int[] arr,int si, int mid,int ei){
        int r = mid+1;
        int count = 0;
        for(int i = si; i <= mid; i++){
            while(r <= ei && (long)arr[i] > 2L*arr[r]){
                r++;
            }
            count += r-(mid+1);
        }
        int n1 = mid - si +1;
        int n2 = ei - mid;
        int[] L = new int[n1];
        int[] R = new int[n2];
        for(int i = 0; i < n1; i++){
            L[i] =arr[si+i];
        }
        for(int i = 0; i < n2; i++){
            R[i] = arr[mid+1+i];
        }
        int i = 0;
        int j = 0;
        int k = si;
        while(i < n1 && j < n2){
            if(L[i] <= R[j]){
                arr[k] = L[i]; 
                i++;
            }else{
                arr[k] = R[j];
                j++;
            }
            k++;
        }
        while(i < n1){
            arr[k] = L[i];
            k++;
            i++;
        }
        while(j < n2){
            arr[k] = R[j];
            j++;
            k++;
        }
        return count;
    }
}